add repodata db support

jmahler opened the pull request:
The repodata in an RPM repo can contain data in both xml and sqlite
formats.  Currently Repodata{Parser,Fetcher} only supports
the xml format.  However, in some cases xml may be missing or it
sqlite might be faster.

Add support for sqlite db formats using new RepodataDb{Parser,Fetcher} classes.

### Testing

Switching to RepodataDb{Fetcher,Parser} works with an existing repo.

```
osdev:repology-updater$ git diff
diff --git a/repos.d/rpm/centos.yaml b/repos.d/rpm/centos.yaml
index f29f2427..0b72f99b 100644
--- a/repos.d/rpm/centos.yaml
+++ b/repos.d/rpm/centos.yaml
@@ -33,8 +33,8 @@
   minpackages: 3000
   sources:
     - name: [ os, updates, extras ]
-      fetcher: RepodataFetcher
-      parser: RepodataParser
+      fetcher: RepodataDbFetcher
+      parser: RepodataDbParser
       url: 'http://vault.centos.org/centos/7/{source}/Source/'
       subrepo: '{source}'
   repolinks:
osdev:repology-updater$ sudo -u repology ./repology-update.py --fetch --fetch --parse --repositories centos_7
Dec 19 16:26:10 loading rules
Dec 19 16:26:13 updating repositories metadata
Dec 19 16:26:13   committing changes
Dec 19 16:26:13 fetching centos_7
Dec 19 16:26:16   done
Dec 19 16:26:27 parsing centos_7
Dec 19 16:26:34   done
Dec 19 16:26:34 total time taken: 23.97 seconds
```

Old Repodata{Fetcher,Parser} still work.

```
osdev:repology-updater$ sudo -u repology ./repology-update.py --fetch --fetch --parse --repositories centos_6
Dec 19 16:28:19 loading rules
Dec 19 16:28:22 updating repositories metadata
Dec 19 16:28:22   committing changes
Dec 19 16:28:22 fetching centos_6
Dec 19 16:28:24   done
Dec 19 16:28:35 parsing centos_6
Dec 19 16:28:42   done
Dec 19 16:28:42 total time taken: 23.45 seconds
```

AMDmi3 commented on repology/fetchers/fetchers/repodata.py:
```diff
@@ -25,21 +25,28 @@
 
         self.url = url
         self.fetch_timeout = fetch_timeout
+        self.use_db = use_db
```
I'd instead set `primary_key = 'primary'` as the class member. Derived class would then override it to `primary_db`. Constructor argument goes away (which is good as it would allow confusing uses), and you don't even need to override constructor in the derived class.

here and remove the constructor argument (which may lead to confusing uses). `RepodataDbFetcher` constructor would set it to `primary_db`

jmahler replied:
Great suggestions, this definitely cleans it up.  Fixing in v2 ...

AMDmi3 commented on repology/fetchers/fetchers/repodata.py:
```diff
@@ -25,21 +25,28 @@
+        self.use_db = use_db
 
     def _do_fetch(self, statefile: AtomicFile, persdata: PersistentData, logger: Logger) -> bool:
+
```
this goes away and later you just use `self.primary_key`

AMDmi3 commented on repology/fetchers/fetchers/repodatadb.py:
```diff
@@ -0,0 +1,24 @@
+from repology.fetchers.fetchers.repodata import RepodataFetcher
+
+
+class RepodataDbFetcher(RepodataFetcher):
```
Let's declare this after the `RepodataFetcher` in the same file. They are logically grouped.

Also I'd name it `RepodataSqliteFetcher` as more meaningful.

AMDmi3 commented on repology/parsers/parsers/repodatadb.py:
```diff
@@ -0,0 +1,47 @@
+from repology.transformer import PackageTransformer
+
+
+class RepodataDbParser(Parser):
```
Likewise, let's move this into the same file with `RepodataParser` and name it `RepodataSqliteParser`.

jmahler commented:
I believe I addressed all the comments.  And code coverage should improve once a source that uses RepodataSqlite{Parser,Fetcher} is added.

Let me know if it needs anything else.

AMDmi3 commented:
I've fixed maintainer extraction (extract_maintainer is needed to get just email) and version handling (same way as it's done in the old parser; interestingly this helped to uncover a bug there). I haven't looked deep into sqlite format, could you give some examples of which information can be available there but not in XML format?
